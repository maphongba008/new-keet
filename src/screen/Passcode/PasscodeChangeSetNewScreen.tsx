import React, { memo, useCallback } from 'react'
import { Text } from 'react-native'

import { IconButton } from 'component/Button'
import { NavBar } from 'component/NavBar'
import SvgIcon from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import { UI_SIZE_24 } from 'lib/commonStyles'
import {
  back,
  navigate,
  SCREEN_PASSCODE_CHANGE_CONFIRM_NEW,
} from 'lib/navigation'

import { useStrings } from 'i18n/strings'

import { PasscodeInput } from './PasscodeInput'

export const PasscodeChangeSetNewScreen: React.FC = () => {
  const strings = useStrings()
  const styles = getStyles()

  const handlePasscodeComplete = useCallback((passcode: string) => {
    navigate(SCREEN_PASSCODE_CHANGE_CONFIRM_NEW, { passcode })
  }, [])

  return (
    <>
      <NavBar
        title={null}
        left={
          <IconButton onPress={back}>
            <SvgIcon name="arrowLeft" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
        middle={
          <Text style={styles.title}>{strings.passcode.changePasscode}</Text>
        }
        right={
          <IconButton onPress={back}>
            <SvgIcon name="close" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
      />
      <PasscodeInput
        title={strings.passcode.inputNew4Digit}
        description={strings.passcode.inputNew4DigitDesc}
        onComplete={handlePasscodeComplete}
        showBiometrics={false}
      />
    </>
  )
}

export default memo(PasscodeChangeSetNewScreen)

const getStyles = createThemedStylesheet(() => ({
  title: {
    color: colors.white_snow,
    textAlign: 'center' as 'center',
  },
}))
