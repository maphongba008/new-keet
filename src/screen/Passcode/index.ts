export * from './PasscodeConfirmationScreen'
export * from './PasscodeInput'
export * from './PasscodeSetupScreen'
