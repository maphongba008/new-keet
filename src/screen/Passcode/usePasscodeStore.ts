import { create } from 'zustand'

export enum MessageType {
  Success = 'success',
  Error = 'error',
}

export enum AutoLockTime {
  OneMinute = '1',
  TwoMinutes = '2',
  FiveMinutes = '5',
  FifteenMinutes = '15',
  Never = 'Never',
}

type MessageTypeOrNull = MessageType | null

interface PasscodeState {
  message: string
  messageType: MessageTypeOrNull
  userHasPasscode: boolean
  autoLockTime: AutoLockTime

  setAutoLockTime: (time: AutoLockTime) => void
  setUserHasPasscode: (value: boolean) => void
  setMessage: (message: string) => void
  setMessageType: (messageType: MessageTypeOrNull) => void
  resetMessage: () => void
}

export const usePasscodeStore = create<PasscodeState>((set) => ({
  message: '',
  messageType: null,
  userHasPasscode: false,
  autoLockTime: AutoLockTime.Never,

  setAutoLockTime: (time: AutoLockTime) => {
    set((state) => {
      if (state.autoLockTime !== time) {
        return { autoLockTime: time }
      }
      return state
    })
  },
  setUserHasPasscode: (value: boolean) => {
    set((state) => {
      if (state.userHasPasscode !== value) {
        return { userHasPasscode: value }
      }
      return state
    })
  },
  setMessage: (message: string) => {
    set((state) => {
      if (state.message !== message) {
        return { message }
      }
      return state
    })
  },
  setMessageType: (messageType: MessageTypeOrNull) => {
    set((state) => {
      if (state.messageType !== messageType) {
        return { messageType }
      }
      return state
    })
  },
  resetMessage: () => set({ message: '', messageType: null }),
}))
