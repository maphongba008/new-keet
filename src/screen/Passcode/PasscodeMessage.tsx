import React, { memo } from 'react'
import { StyleSheet, Text } from 'react-native'

import { createThemedStylesheet } from 'component/theme'
import { UI_SIZE_8 } from 'lib/commonStyles'

import { MessageType } from './usePasscodeStore'

type PasscodeMessageProps = {
  message: string
  messageType: MessageType | null
}

const PasscodeMessage: React.FC<PasscodeMessageProps> = ({
  message,
  messageType,
}) => {
  const styles = getStyles()

  if (!message) return null

  return (
    <Text
      style={
        messageType === MessageType.Success
          ? styles.successMessage
          : styles.errorMessage
      }
    >
      {message}
    </Text>
  )
}

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    errorMessage: {
      color: theme.color.red_400,
      marginTop: UI_SIZE_8,
      textAlign: 'center',
    },
    successMessage: {
      color: theme.color.green_300,
      marginTop: UI_SIZE_8,
      textAlign: 'center',
    },
  })

  return styles
})

export default memo(PasscodeMessage)
