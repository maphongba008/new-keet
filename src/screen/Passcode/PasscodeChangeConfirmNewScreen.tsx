import React, { memo, useCallback, useEffect } from 'react'
import { Text } from 'react-native'
import { useRoute } from '@react-navigation/native'

import { IconButton } from 'component/Button'
import { NavBar } from 'component/NavBar'
import SvgIcon from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import { UI_SIZE_24 } from 'lib/commonStyles'
import { useTimeout } from 'lib/hooks'
import { setStoragePasscode } from 'lib/localStorage'
import { APP_ROOT, back, navigate } from 'lib/navigation'

import { useStrings } from 'i18n/strings'

import { PasscodeInput } from './PasscodeInput'
import PasscodeMessage from './PasscodeMessage'
import { MessageType, usePasscodeStore } from './usePasscodeStore'

const PasscodeChangeConfirmNewScreen: React.FC = () => {
  const strings = useStrings()
  const styles = getStyles()
  const route = useRoute()
  const { setUserHasPasscode } = usePasscodeStore()
  const { setMessage, setMessageType, message, messageType, resetMessage } =
    usePasscodeStore()

  const newPasscode = (route.params as { passcode: string }).passcode

  const navigateToRootWithDelay = useTimeout(() => navigate(APP_ROOT), 1000)

  const handlePasscodeComplete = useCallback(
    (confirmedPasscode: string) => {
      if (confirmedPasscode === newPasscode) {
        setStoragePasscode(confirmedPasscode)
        setMessageType(MessageType.Success)
        setUserHasPasscode(true)
        setMessage(strings.passcode.passcodeSettingSuccessfull)
        navigateToRootWithDelay()
      } else {
        setMessageType(MessageType.Error)
        setMessage(strings.passcode.passcodeDoesNotMatch)
      }
    },
    [
      newPasscode,
      setMessage,
      setMessageType,
      setUserHasPasscode,
      strings.passcode.passcodeDoesNotMatch,
      strings.passcode.passcodeSettingSuccessfull,
      navigateToRootWithDelay,
    ],
  )

  useEffect(() => {
    resetMessage()
  }, [resetMessage])

  return (
    <>
      <NavBar
        title={null}
        left={
          <IconButton onPress={back}>
            <SvgIcon name="arrowLeft" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
        middle={
          <Text style={styles.title}>{strings.passcode.changePasscode}</Text>
        }
        right={
          <IconButton onPress={back}>
            <SvgIcon name="close" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
      />
      <PasscodeInput
        title={strings.passcode.confirmNew4Digit}
        description={strings.passcode.confirmNew4DigitDesc}
        onComplete={handlePasscodeComplete}
        showBiometrics={false}
      />
      <PasscodeMessage message={message} messageType={messageType} />
    </>
  )
}

export default memo(PasscodeChangeConfirmNewScreen)

const getStyles = createThemedStylesheet(() => ({
  title: {
    color: colors.white_snow,
    textAlign: 'center' as 'center',
  },
}))
