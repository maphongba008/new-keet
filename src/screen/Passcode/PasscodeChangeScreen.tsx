import React, { memo, useCallback, useEffect } from 'react'
import { Text } from 'react-native'

import { IconButton } from 'component/Button'
import { NavBar } from 'component/NavBar'
import SvgIcon from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import { UI_SIZE_24 } from 'lib/commonStyles'
import { getStoragePasscode } from 'lib/localStorage'
import { back, navigate, SCREEN_PASSCODE_CHANGE_SET_NEW } from 'lib/navigation'

import { useStrings } from 'i18n/strings'

import { PasscodeInput } from './PasscodeInput'
import PasscodeMessage from './PasscodeMessage'
import { MessageType, usePasscodeStore } from './usePasscodeStore'

const PasscodeChangeScreen: React.FC = () => {
  const strings = useStrings()
  const styles = getStyles()
  const { setMessage, setMessageType, message, messageType, resetMessage } =
    usePasscodeStore()

  const handlePasscodeComplete = useCallback(
    (passcode: string) => {
      const currentPasscode = getStoragePasscode()

      if (passcode === currentPasscode) {
        navigate(SCREEN_PASSCODE_CHANGE_SET_NEW, { passcode })
        setMessageType(MessageType.Success)
        setMessage(strings.passcode.passcodeCorrect)
      } else {
        setMessageType(MessageType.Error)
        setMessage(strings.passcode.passcodeIncorrect)
      }
    },
    [
      setMessage,
      setMessageType,
      strings.passcode.passcodeCorrect,
      strings.passcode.passcodeIncorrect,
    ],
  )

  useEffect(() => {
    resetMessage()

    return () => {
      resetMessage()
    }
  }, [resetMessage])

  return (
    <>
      <NavBar
        title={null}
        left={
          <IconButton onPress={back}>
            <SvgIcon name="arrowLeft" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
        middle={
          <Text style={styles.title}>{strings.passcode.changePasscode}</Text>
        }
        right={
          <IconButton onPress={back}>
            <SvgIcon name="close" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
      />
      <PasscodeInput
        title={strings.passcode.inputOld4Digit}
        description={strings.passcode.inputOld4DigitDesc}
        onComplete={handlePasscodeComplete}
        showBiometrics={false}
      />
      <PasscodeMessage message={message} messageType={messageType} />
    </>
  )
}

export default memo(PasscodeChangeScreen)

const getStyles = createThemedStylesheet(() => ({
  title: {
    color: colors.white_snow,
    textAlign: 'center' as 'center',
  },
}))
