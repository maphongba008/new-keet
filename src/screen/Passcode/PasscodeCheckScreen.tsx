import React, { memo, useCallback, useEffect } from 'react'
import { Text, TextStyle, View } from 'react-native'

import { IconButton } from 'component/Button'
import { NavBar } from 'component/NavBar'
import SvgIcon from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import { UI_SIZE_24 } from 'lib/commonStyles'
import { useAppStore, useTimeout } from 'lib/hooks'
import { getStoragePasscode } from 'lib/localStorage'
import {
  back,
  getCurrentRoute,
  reset,
  SCREEN_HOME,
  SCREEN_PASSCODE_CHECK,
} from 'lib/navigation'

// Adjust the import path as necessary

import { useStrings } from 'i18n/strings'

import { BIOMETRICS_SUCCESS, PasscodeInput } from './PasscodeInput'
import PasscodeMessage from './PasscodeMessage'
import { MessageType, usePasscodeStore } from './usePasscodeStore'

interface PasscodeCheckProps {
  title?: string
  description?: string
  titleStyleProp?: TextStyle
  onSuccess?: () => void
  hasBackButton?: boolean
  hasCloseButton?: boolean
}

const PasscodeCheckScreen: React.FC<PasscodeCheckProps> = ({
  title,
  description,
  titleStyleProp,
  onSuccess,
  hasBackButton = false,
  hasCloseButton = false,
}) => {
  const strings = useStrings()
  const styles = getStyles()
  const { message, messageType, setMessage, setMessageType, resetMessage } =
    usePasscodeStore()

  const handleCorrectPasscodeWithDelay = useTimeout(() => {
    useAppStore.setState({
      isAuthenticated: true,
    })
    if (onSuccess) {
      onSuccess()
    } else {
      const currentRoute = getCurrentRoute()
      if (currentRoute === SCREEN_PASSCODE_CHECK) {
        back()
      } else {
        reset(SCREEN_HOME)
      }
    }
  }, 500)

  const handlePasscodeComplete = useCallback(
    (passcode: string) => {
      const isCorrectPasscode =
        passcode === BIOMETRICS_SUCCESS || passcode === getStoragePasscode()

      if (isCorrectPasscode) {
        setMessageType(MessageType.Success)
        setMessage(strings.passcode.passcodeCorrect)
        handleCorrectPasscodeWithDelay()
      } else {
        setMessageType(MessageType.Error)
        setMessage(strings.passcode.passcodeDoesNotMatch)
      }
    },
    [
      setMessage,
      setMessageType,
      strings.passcode.passcodeCorrect,
      strings.passcode.passcodeDoesNotMatch,
      handleCorrectPasscodeWithDelay,
    ],
  )
  useEffect(() => {
    resetMessage()
  }, [resetMessage])

  return (
    <>
      <NavBar
        title={null}
        left={
          hasBackButton ? (
            <IconButton onPress={back}>
              <SvgIcon
                name="arrowLeft"
                width={UI_SIZE_24}
                height={UI_SIZE_24}
              />
            </IconButton>
          ) : null
        }
        middle={
          !title ? (
            <Text style={styles.title}>{strings.passcode.inputPasscode}</Text>
          ) : null
        }
        right={
          hasCloseButton ? (
            <IconButton onPress={back}>
              <SvgIcon name="close" width={UI_SIZE_24} height={UI_SIZE_24} />
            </IconButton>
          ) : null
        }
      />
      <View style={styles.container}>
        <PasscodeInput
          titleStyleProp={titleStyleProp}
          title={title ?? strings.passcode.input4DigitPasscode}
          description={description ?? strings.passcode.inputPasscodeDesc}
          onComplete={handlePasscodeComplete}
        />
        <PasscodeMessage message={message} messageType={messageType} />
      </View>
    </>
  )
}

export default memo(PasscodeCheckScreen)

const getStyles = createThemedStylesheet((theme) => ({
  container: {
    alignItems: 'center' as 'center',
    backgroundColor: theme.background.bg_1,
    flex: 1,
  },
  title: {
    color: colors.white_snow,
    textAlign: 'center' as 'center',
  },
}))
