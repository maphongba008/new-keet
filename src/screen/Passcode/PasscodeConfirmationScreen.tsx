import React, { memo, useCallback } from 'react'
import { Text } from 'react-native'
import { useRoute } from '@react-navigation/native'

import { IconButton } from 'component/Button'
import { NavBar } from 'component/NavBar'
import SvgIcon from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import { APPIUM_IDs, appiumTestProps } from 'lib/appium'
import { UI_SIZE_24 } from 'lib/commonStyles'
import { useTimeout } from 'lib/hooks'
import { setStoragePasscode } from 'lib/localStorage'
import { APP_ROOT, back, navigate } from 'lib/navigation'

import { useStrings } from 'i18n/strings'

import { PasscodeInput } from './PasscodeInput'
import PasscodeMessage from './PasscodeMessage'
import { MessageType, usePasscodeStore } from './usePasscodeStore'

const PasscodeConfirmationScreen: React.FC = () => {
  const strings = useStrings()
  const route = useRoute()
  const styles = getStyles()
  const {
    setMessage,
    setMessageType,
    message,
    messageType,
    setUserHasPasscode,
  } = usePasscodeStore()

  const initialPasscode = (route.params as { passcode: string }).passcode

  const navigateToRootWithDelay = useTimeout(() => navigate(APP_ROOT), 1000)

  const handlePasscodeComplete = useCallback(
    (confirmedPasscode: string) => {
      if (confirmedPasscode === initialPasscode) {
        setStoragePasscode(confirmedPasscode)
        setMessageType(MessageType.Success)
        setMessage(strings.passcode.passcodeSettingSuccessfull)
        setUserHasPasscode(true)
        navigateToRootWithDelay()
      } else {
        setMessageType(MessageType.Error)
        setMessage(strings.passcode.passcodeDoesNotMatch)
      }
    },
    [
      initialPasscode,
      setMessage,
      setMessageType,
      setUserHasPasscode,
      strings.passcode.passcodeDoesNotMatch,
      strings.passcode.passcodeSettingSuccessfull,
      navigateToRootWithDelay,
    ],
  )

  return (
    <>
      <NavBar
        title={null}
        left={
          <IconButton
            onPress={back}
            {...appiumTestProps(APPIUM_IDs.passcode_confirm_btn_back)}
          >
            <SvgIcon name="arrowLeft" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
        middle={
          <Text style={styles.title}>{strings.passcode.setupPasscode}</Text>
        }
        right={
          <IconButton
            onPress={back}
            {...appiumTestProps(APPIUM_IDs.passcode_confirm_btn_close)}
          >
            <SvgIcon name="close" width={UI_SIZE_24} height={UI_SIZE_24} />
          </IconButton>
        }
      />
      <PasscodeInput
        title={strings.passcode.confirm4Digit}
        description={strings.passcode.confirm4DigitDesc}
        onComplete={handlePasscodeComplete}
        showBiometrics={false}
      />
      <PasscodeMessage message={message} messageType={messageType} />
    </>
  )
}

export default memo(PasscodeConfirmationScreen)

const getStyles = createThemedStylesheet(() => ({
  title: {
    color: colors.white_snow,
    textAlign: 'center' as 'center',
  },
}))
