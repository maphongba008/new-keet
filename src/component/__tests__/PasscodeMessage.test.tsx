import React from 'react'
import { render } from '@testing-library/react-native'

import PasscodeMessage from 'screen/Passcode/PasscodeMessage'
import { MessageType } from 'screen/Passcode/usePasscodeStore'

describe('PasscodeMessage Snapshot', () => {
  it('should match the snapshot with a success message', () => {
    const { toJSON } = render(
      <PasscodeMessage
        message="Passcode correct"
        messageType={MessageType.Success}
      />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should match the snapshot with an error message', () => {
    const { toJSON } = render(
      <PasscodeMessage
        message="Passcode incorrect"
        messageType={MessageType.Error}
      />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should match the snapshot with no message', () => {
    const { toJSON } = render(<PasscodeMessage message="" messageType={null} />)
    expect(toJSON()).toMatchSnapshot()
  })
})
