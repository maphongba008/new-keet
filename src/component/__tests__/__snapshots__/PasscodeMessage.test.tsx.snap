// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasscodeMessage Snapshot should match the snapshot with a success message 1`] = `
<Text
  style={
    {
      "color": "#53EA98",
      "marginTop": 8,
      "textAlign": "center",
    }
  }
>
  Passcode correct
</Text>
`;

exports[`PasscodeMessage Snapshot should match the snapshot with an error message 1`] = `
<Text
  style={
    {
      "color": "#FF6D6D",
      "marginTop": 8,
      "textAlign": "center",
    }
  }
>
  Passcode incorrect
</Text>
`;

exports[`PasscodeMessage Snapshot should match the snapshot with no message 1`] = `null`;
